src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/PosixSignalRegistration.Unix.cs

```diff
@@ -0,0 +1,144 @@
+        {
+            _signal = signal;
+            _handler = handler;
+        }
+
+        public static PosixSignalRegistration Create(PosixSignal signal, Action<PosixSignalContext> handler)
+        {
+            if (handler == null)
+            {
+                throw new ArgumentNullException(nameof(handler));
+            }
+            if (signal > PosixSignal.SIGHUP || signal < PosixSignal.SIGCHLD)
+            {
+                throw new IndexOutOfRangeException();
+            }
+            PosixSignalRegistration registration = new PosixSignalRegistration(signal, handler);
+            registration.RegisterFor(signal);
+            return registration;
+        }
+
+        private unsafe void RegisterFor(PosixSignal signal)
+        {
+            if (!s_initialized)
+            {
+                if (!Interop.Sys.InitializeTerminalAndSignalHandling())
+                {
+                    throw new Win32Exception();
+                }
+                s_initialized = true;
+            }
+            lock (s_registrations)
+            {
+                if (!s_registrations.TryGetValue(signal, out List<PosixSignalRegistration>? signalRegistrations))
+                {
+                    if (!Interop.Sys.RegisterForPosixSignal(signal, &OnPosixSignal))
+                    {
+                        throw new Win32Exception();
+                    }
+                    signalRegistrations = new List<PosixSignalRegistration>();
+                    s_registrations.Add(signal, signalRegistrations);
+                }
+                signalRegistrations.Add(this);
+            }
+            _registered = true;
+        }
+
+        private void Handle(PosixSignalContext context)
+        {
+            lock (_gate)
+            {
+                if (_registered)
+                {
+                    _handler(context);
+                }
+            }
+        }
+
+        [UnmanagedCallersOnly]
+        private static int OnPosixSignal(PosixSignal signal)
+        {
+            lock (s_registrations)
+            {
+                if (s_registrations.TryGetValue(signal, out List<PosixSignalRegistration>? signalRegistrations))
+                {
+                    if (signalRegistrations.Count != 0)
+                    {
+                        PosixSignalRegistration[] registrations = signalRegistrations.ToArray();
+                        // This is called on the native signal handling thread. We need to move to another thread so
+                        // signal handling is not blocked. Otherwise we may get deadlocked when the handler depends
+                        // on work triggered from the signal handling thread.
+                        // We use a new thread rather than queueing to the ThreadPool in order to prioritize handling
+                        // in case the ThreadPool is saturated.
+                        Thread handlerThread = new Thread(HandleSignal)
```

Comment:
Maybe we want to defer some signals to the `ThreadPool` instead of starting a new `Thread`. From the current supported signals, maybe `SIGCHLD`?

The other signals are meant for termination, so we want prioritize them.

Reply:
Maybe the ones that fire more often yeah.

Reply:
If the implementation is going to prioritize different signals differently, I think that should be an option on the `PosixSignalRegistration`. I don't think the runtime knows better than users which signals are going to fire how often in any number of possible scenarios.

Reply:
I've updated the implementation to only start a new `Thread` for a specific set of signals.
Using the `ThreadPool` by default should be fine.